### PR TITLE
Small changes to OCI repo client

### PR DESCRIPTION
**Pull request by rylev**

### CODE REVIEW

File: crates/oci-distribution/src/lib.rs [hunk 1]
```diff
@@ -63,37 +62,36 @@ impl Client {
...
-        }
-        let auth = WwwAuthenticate::parse_header(&disthdr.unwrap().as_bytes().into())?;
-        let challenge_opt = auth.get::<BearerChallenge>();
-        if challenge_opt.is_none() {
-            // This means that no challenge was present, even though the header was present.
-            // Since we do not handle basic auth, it could be the case that the upstream service
-            // is in compatibility mode with a Docker v1 registry.
-            return Ok(());
-        }
+        let dist_hdr = match res.headers().get(reqwest::header::WWW_AUTHENTICATE) {
```
Comment (technosophos) [inline crates/oci-distribution/src/lib.rs]: Oh! I see! This is a great way of doing things.

File: crates/oci-distribution/src/lib.rs [hunk 2]
```diff
@@ -129,34 +126,29 @@ impl Client {
...
-            // According to the OCI spec, we should see an error in the message body.
-            let err = res.json::<OciEnvelope>().await?;
-            // FIXME: This should not have to wrap the error.
-            Err(format_err!("{} on {}", err.errors[0], url))
-        } else if status.is_server_error() {
-            Err(format_err!("Server error at {}", url))
-        } else {
-            let text = res.text().await?;
-            Err(format_err!(
+        match res.status() {
```
Comment (technosophos) [inline crates/oci-distribution/src/lib.rs]: Ooo! I like this, too! I definitely have not gotten the hang of the match/if syntax.

File: crates/oci-distribution/src/lib.rs [hunk 3]
```diff
@@ -201,14 +183,18 @@ impl Challenge for BearerChallenge {
...
+
     fn into_raw(self) -> RawChallenge {
         let mut map = ChallengeFields::new();
-        self.realm
-            .and_then(|realm| map.insert_static_quoting("realm", realm));
-        self.scope
-            .and_then(|item| map.insert_static_quoting("scope", item));
-        self.service
-            .and_then(|item| map.insert_static_quoting("service", item));
+        if let Some(realm) = self.realm {
```
Comment (technosophos) [inline crates/oci-distribution/src/lib.rs]: Okay, I have to say that I find the if/map thing less readable than the and_then() thing. Is this an established pattern or a preference?

File: crates/oci-distribution/Cargo.toml [hunk 4]
```diff
@@ -16,7 +16,7 @@ edition = "2018"
 reqwest = { version = "0.10.4", features = ["json"] }
 failure = "0.1"
 tokio = "0.2.13"
-serde = "1.0"
+serde = { version = "1.0", features = ["derive"] }
```
Comment (technosophos) [inline crates/oci-distribution/Cargo.toml]: Still trying to learn about how features work... does this mean "the feature `derive` is added" or "all the features except `derive` are disabled"?

File: crates/oci-distribution/src/lib.rs (hunk 3 above)
Comment (rylev) [reply crates/oci-distribution/src/lib.rs]: I would say using `if let` is much more common than `and_then` in this case. `and_then` is really best when wanting to do the same as a `map` then `flatten`. We simply want to perform an action if the item is set.

File: crates/oci-distribution/Cargo.toml (hunk 4 above)
Comment (rylev) [reply crates/oci-distribution/Cargo.toml]: Features are generally additive (i.e., they _should_ be additive but not every crate follows that recommended pattern). So putting the `derive` feature here adds additional functionality (namely the ability to derive `Serialize` and `Deserialize`). You may see `default-features = false` which turns off all features that are marked as default in that dependency's Cargo.toml file. You can then add features on top of that.

**Pull request merged**